helix-core/src/syntax.rs

```diff
@@ -857,6 +868,12 @@ impl Syntax {
                     }
                 }
             }
+
+            let hash = layers_hasher.hash_one(layer);
+            // Safety: insert_no_grow is unsafe because it assumes that the table
+            // has enough capacity to hold additional elements.
+            // This is always the case as we reserved enough capacity above.
+            unsafe { layers_table.insert_no_grow(hash, layer_id) };
```

Comment:
Why not just `insert()`? It's guaranteed to avoid allocation if there's enough capacity, but it won't panic if that's not the case

Reply:
I usually use `insert_no_grow` with `RawTable` if the capacity is trivially large enough for two reasons:
* It's faster: Swiss tables need to check an additional byte to determine if a bucket is empty so the overhead is larger then for a `Vec` because you need to read from a random location in the table. The complexity of this bounds check also means its basically never optimized out.
* `insert` requires a closure for rehashing existing entries in case the table needs to be resized (which is impossible because we have enough capacity). That either adds a unnecessary colsure that calculates a hash or a weird looking `unrechable` as shown below so I usually prefer the `insert_no_grow` variant over the other options:
``` rust
layers_table.insert(hash, layer_id, |_| unreachable!("table should be large enough"));
layers_table.insert(hash, layer_id, |rehashed_layer| hasher.hash_one(self.layers[rehashed_layer])));
```